<center>
<h4>Diplomatura en CDAAyA 2018 - FaMAF - UNC</h4>
<h2>Laboratorio de Exploración y Curación de Datos</h2>
<h3>Limpieza de Base de Permisos de Cosntrucción en San Francisco</h3>
</center>
</left>
<h4>Lucas Prone y Julieta Bergamasco</h4>
</left>

## Introducción

En la siguiente notebook, implementaremos los pasos necesarios para la limpieza de una base de datos de permisos de construcción en San Francisco, EEUU. Para ello, seguiremos el *check list* propuesto en clase.

## 1. Importación de Datos

### 1.1. Verificación de Inexistencia de Problemas en la Importación

Para comenzar, importamos los datos que vamos a procesar:

In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('Building_Permits.csv',low_memory=False)

In [3]:
pd.set_option('display.max_columns', None)
data.head(5)

Permit Number  Permit Type            Permit Type Definition  \
0  201505065519            4                      sign - erect   
1  201604195146            4                      sign - erect   
2  201605278609            3  additions alterations or repairs   
3  201611072166            8            otc alterations permit   
4  201611283529            6                       demolitions   

  Permit Creation Date Block  Lot  Street Number Street Number Suffix  \
0           05/06/2015  0326  023            140                  NaN   
1           04/19/2016  0306  007            440                  NaN   
2           05/27/2016  0595  203           1647                  NaN   
3           11/07/2016  0156  011           1230                  NaN   
4           11/28/2016  0342  001            950                  NaN   

  Street Name Street Suffix  Unit Unit Suffix  \
0       Ellis            St   NaN         NaN   
1       Geary            St   0.0         NaN   
2     Pacific            Av   NaN         NaN   
3     Pacific            Av   0.0         NaN   
4      Market            St   NaN         NaN   

                                         Description Current Status  \
0  ground fl facade: to erect illuminated, electr...        expired   
1            remove (e) awning and associated signs.         issued   
2                    installation of separating wall      withdrawn   
3           repair dryrot & stucco at front of bldg.       complete   
4  demolish retail/office/commercial 3-story buil...         issued   

  Current Status Date  Filed Date Issued Date Completed Date  \
0          12/21/2017  05/06/2015  11/09/2015            NaN   
1          08/03/2017  04/19/2016  08/03/2017            NaN   
2          09/26/2017  05/27/2016         NaN            NaN   
3          07/24/2017  11/07/2016  07/18/2017     07/24/2017   
4          12/01/2017  11/28/2016  12/01/2017            NaN   

  First Construction Document Date Structural Notification  \
0                       11/09/2015                     NaN   
1                       08/03/2017                     NaN   
2                              NaN                     NaN   
3                       07/18/2017                     NaN   
4                       11/20/2017                     NaN   

   Number of Existing Stories  Number of Proposed Stories  \
0                         6.0                         NaN   
1                         7.0                         NaN   
2                         6.0                         6.0   
3                         2.0                         2.0   
4                         3.0                         NaN   

  Voluntary Soft-Story Retrofit Fire Only Permit Permit Expiration Date  \
0                           NaN              NaN             11/03/2016   
1                           NaN              NaN             12/03/2017   
2                           NaN              NaN                    NaN   
3                           NaN              NaN             07/13/2018   
4                           NaN              NaN             12/01/2018   

   Estimated Cost  Revised Cost         Existing Use  Existing Units  \
0          4000.0        4000.0  tourist hotel/motel           143.0   
1             1.0         500.0  tourist hotel/motel             NaN   
2         20000.0           NaN         retail sales            39.0   
3          2000.0        2000.0    1 family dwelling             1.0   
4        100000.0      100000.0         retail sales             NaN   

        Proposed Use  Proposed Units  Plansets TIDF Compliance  \
0                NaN             NaN       2.0             NaN   
1                NaN             NaN       2.0             NaN   
2       retail sales            39.0       2.0             NaN   
3  1 family dwelling             1.0       2.0             NaN   
4                NaN             NaN       2.0             NaN   

   Existing Construction Type Existing

Tomamos una muestra aleatoria para ver valores más dispersos.

In [4]:
# set seed for reproducibility
import numpy as np
np.random.seed(0)
data.sample(20)

Permit Number  Permit Type            Permit Type Definition  \
40553   201403039652            8            otc alterations permit   
169731  201510159735            3  additions alterations or repairs   
19180        M409787            8            otc alterations permit   
68047   201411191888            8            otc alterations permit   
64238        M527228            8            otc alterations permit   
27703        M434527            8            otc alterations permit   
109281  201512175360            8            otc alterations permit   
15075   201306109020            8            otc alterations permit   
59124   201408264765            8            otc alterations permit   
175167  201702240086            8            otc alterations permit   
157845  201705247586            8            otc alterations permit   
150068       M762087            8            otc alterations permit   
169291  201708164983            8            otc alterations permit   
61608   201409196774            3  additions alterations or repairs   
34145   201312174422            8            otc alterations permit   
181163  201709067013            8            otc alterations permit   
593          M364468            8            otc alterations permit   
198604  201802221963            8            otc alterations permit   
117133       M671487            8            otc alterations permit   
166700       M815987            8            otc alterations permit   

       Permit Creation Date Block   Lot  Street Number Street Number Suffix  \
40553            03/03/2014  3732   008            400                  NaN   
169731           10/15/2015  2609   028             79                  NaN   
19180            07/22/2013  4624   031            178                  NaN   
68047            11/19/2014  0039   109            294                  NaN   
64238            10/14/2014  1251   002            707                  NaN   
27703            10/11/2013  0348   015            277                  NaN   
109281           12/17/2015  0061   013             40                  NaN   
15075            06/10/2013  0262   020            100                  NaN   
59124            08/26/2014  0321   016            650                  NaN   
175167           02/24/2017  6533   028            451                  NaN   
157845           05/24/2017  2630   018           1171                  NaN   
150068           02/17/2017  1776  009D           1355                  NaN   
169291           08/16/2017  2347  004G           2350                  NaN   
61608            09/19/2014  2705   029             98                  NaN   
34145            12/17/2013  3514   030           1663                  NaN   
181163           09/06/2017  0289   005            100                  NaN   
593              01/03/2013  0604   023           2330                  NaN   
198604           02/22/2018  0931   013           2544                  NaN   
117133           03/10/2016  5651   027            113                  NaN   
166700           08/03/2017  0868   066            108                    B   

        Street Name Street Suffix  Unit Unit Suffix  \
40553    Clementina            St   NaN         NaN   
169731  Buena Vista            Tr   NaN         NaN   
19180    West Point            Rd   NaN         NaN   
68047     Francisco            St   0.0         NaN   
64238          Cole            St   NaN         NaN   
27703   Golden Gate            Av   0.0         NaN   
109281      Whiting            St   NaN         NaN   
15075          Pine            St   0.0         NaN   
59124         Ellis            St   NaN         NaN   
175167    Fair Oaks            St   NaN         NaN   
157845      Stanyan            St   NaN         NaN   
150068         21st            Av   NaN         NaN   
169291         19th            Av   NaN         NaN   
61608         Crown            Tr   NaN         NaN   
34145       Mission           

In [5]:
data.iloc[:,0:15].sample(20)

Permit Number  Permit Type            Permit Type Definition  \
60635   201409105988            3  additions alterations or repairs   
110430  201601046309            8            otc alterations permit   
165042  201707212672            3  additions alterations or repairs   
79200   201503171100            8            otc alterations permit   
76487   201502208860            8            otc alterations permit   
103125  201510210411            8            otc alterations permit   
109832  201512235838            8            otc alterations permit   
47606   201405125430            8            otc alterations permit   
87148   201505297472            8            otc alterations permit   
35831   201401085824            8            otc alterations permit   
194652       M877448            8            otc alterations permit   
60894   201409126195            8            otc alterations permit   
12540   201305146914            8            otc alterations permit   
173889  201610180598            8            otc alterations permit   
68458   201411212247            3  additions alterations or repairs   
77102        M563307            8            otc alterations permit   
55830   201407252156            8            otc alterations permit   
68433   201411212226            3  additions alterations or repairs   
143674  201611293663            8            otc alterations permit   
40995   201403060031            8            otc alterations permit   

       Permit Creation Date  Block   Lot  Street Number Street Number Suffix  \
60635            09/10/2014   0479   027           1356                  NaN   
110430           01/04/2016   1276  001H            202                  NaN   
165042           07/21/2017   3622   004            818                  NaN   
79200            03/17/2015   1509   009            547                  NaN   
76487            02/20/2015   1923   023           1662                  NaN   
103125           10/21/2015   0738   049            851                  NaN   
109832           12/23/2015   0818   046            488                  NaN   
47606            05/12/2014   5523   007           3240                    A   
87148            05/29/2015  2823A   017              5                  NaN   
35831            01/08/2014   1857   030           1526                  NaN   
194652           01/17/2018   0770   015            916                  NaN   
60894            09/12/2014   2612   086            261                  NaN   
12540            05/14/2013   3525   107            465                  NaN   
173889           10/18/2016   5319   022           1982                  NaN   
68458            11/21/2014   3598   028            655                  NaN   
77102            02/25/2015   3755   108            158                  NaN   
55830            07/25/2014   0216   005           1315                  NaN   
68433            11/21/2014  2901A   050             15                  NaN   
143674           11/29/2016   3719   018            199                  NaN   
40995            03/06/2014   5308  017A           1445                  NaN   

        Street Name Street Suffix   Unit Unit Suffix  \
60635      Chestnut            St    0.0         NaN   
110430      Grattan            St    0.0         NaN   
165042          Noe            St    NaN         NaN   
79200          35th            Av    NaN         NaN   
76487          21st            Av    NaN         NaN   
103125     Van Ness            Av  101.0         NaN   
109832         Fell            St    NaN         NaN   
47606        Folsom            St    NaN         NaN   
87148     Gladeview            Wy    NaN         NaN   
35831          12th            Av    NaN         NaN   
194652       Laguna            St    NaN         NaN   
60894         Henry            St    NaN         NaN   
12540          10th            St  304.0         NaN   
173889        Palou            Av    NaN         N

In [6]:
data.iloc[:,15:23].sample(20)

Filed Date Issued Date Completed Date  \
155499  04/26/2017  04/26/2017            NaN   
39974   02/24/2014  02/24/2014     04/24/2014   
174870  01/30/2017  01/30/2017     01/11/2018   
143712  11/29/2016  11/29/2016            NaN   
65453   10/24/2014  12/05/2014     04/30/2015   
32697   12/03/2013  12/03/2013            NaN   
14118   05/30/2013  06/13/2013     04/15/2014   
33837   12/13/2013  12/13/2013            NaN   
66012   10/30/2014  10/30/2014            NaN   
119639  04/01/2016  04/08/2016     12/29/2016   
41266   03/10/2014  03/10/2014            NaN   
90162   06/25/2015  07/02/2015            NaN   
136790  09/15/2016  04/04/2017            NaN   
124067  05/13/2016  03/21/2017            NaN   
117176  03/10/2016  06/01/2017            NaN   
159512  06/15/2017         NaN            NaN   
4689    02/19/2013  02/19/2013            NaN   
55780   07/24/2014  02/19/2015            NaN   
49647   05/29/2014  05/29/2014            NaN   
11220   04/30/2013         NaN            NaN   

       First Construction Document Date Structural Notification  \
155499                       04/26/2017                     NaN   
39974                        02/24/2014                     NaN   
174870                       01/30/2017                     NaN   
143712                       11/29/2016                     NaN   
65453                        12/05/2014                     NaN   
32697                        12/03/2013                     NaN   
14118                        06/13/2013                     NaN   
33837                        12/13/2013                     NaN   
66012                        10/30/2014                     NaN   
119639                       04/08/2016                     NaN   
41266                        03/10/2014                     NaN   
90162                        07/02/2015                     NaN   
136790                       04/04/2017                     NaN   
124067                       03/21/2017                       Y   
117176                       06/01/2017                     NaN   
159512                              NaN                     NaN   
4689                         02/19/2013                     NaN   
55780                        02/19/2015                     NaN   
49647                        05/29/2014                     NaN   
11220                               NaN                     NaN   

        Number of Existing Stories  Number of Proposed Stories  \
155499                         NaN                         NaN   
39974                          3.0                         3.0   
174870                         3.0                         3.0   
143712                         NaN                         NaN   
65453                          3.0                         3.0   
32697                          2.0                         2.0   
14118                          3.0                         3.0   
33837                          2.0                         3.0   
66012                          NaN                         NaN   
119639                         2.0                         2.0   
41266                         29.0                        29.0   
90162                          1.0                         1.0   
136790                         3.0                         3.0   
124067                         3.0                         3.0   
117176                         3.0                         3.0   
159512                         4.0                         4.0   
4689                           NaN                         NaN   
55780                          2.0                         NaN   
49647                          NaN                         NaN   
11220                          2.0                         2.0   

       Voluntary Soft-Story Retrofit  
155499                           NaN  
39974                            NaN  
174870                           NaN  
143712                     

In [7]:
data.iloc[:,23:33].sample(20)

Fire Only Permit Permit Expiration Date  Estimated Cost  Revised Cost  \
863                 NaN                    NaN             NaN           1.0   
15651               NaN             06/14/2014         30000.0       40000.0   
102534                Y             10/10/2016         12000.0       12000.0   
52341               NaN             06/19/2015         15000.0       15000.0   
83758               NaN             04/21/2016         18000.0       36000.0   
100819              NaN                    NaN             NaN           1.0   
170983              NaN                    NaN         47000.0           0.0   
123601              NaN             05/04/2017          4000.0       12700.0   
29728               NaN             11/04/2016        170000.0      170000.0   
26544               NaN             09/26/2014             1.0           1.0   
54261               NaN                    NaN             NaN           1.0   
124928              NaN                    NaN             NaN           1.0   
171373              NaN             11/18/2018        140000.0      140000.0   
32076               NaN             11/22/2014         25000.0       35000.0   
194239                Y             05/19/2018             1.0           1.0   
41608               NaN             03/11/2017        750000.0     1160000.0   
60570               NaN                    NaN             NaN           1.0   
46645               NaN                    NaN             NaN           1.0   
40226               NaN             02/21/2015         14780.0       14780.0   
95790               NaN             08/12/2016          1200.0        1200.0   

                Existing Use  Existing Units       Proposed Use  \
863                      NaN             NaN                NaN   
15651      1 family dwelling             1.0  1 family dwelling   
102534                office             0.0             office   
52341                 office             0.0             office   
83758             apartments            15.0         apartments   
100819                   NaN             NaN                NaN   
170983            apartments             5.0         apartments   
123601            apartments             3.0         apartments   
29728      1 family dwelling             1.0  1 family dwelling   
26544      2 family dwelling             2.0  2 family dwelling   
54261                    NaN             NaN                NaN   
124928                   NaN             NaN                NaN   
171373  warehouse,no frnitur             2.0         greenhouse   
32076             apartments            10.0         apartments   
194239            apartments           100.0         apartments   
41608                 office            33.0             office   
60570                    NaN             NaN                NaN   
46645                    NaN             NaN                NaN   
40226             apartments            15.0         apartments   
95790      2 family dwelling             2.0  2 family dwelling   

        Proposed Units  Plansets TIDF Compliance  
863                NaN       NaN             NaN  
15651              1.0       2.0             NaN  
102534             0.0       2.0             NaN  
52341              0.0       2.0             NaN  
83758             15.0       0.0             NaN  
100819             NaN       NaN             NaN  
170983             5.0       2.0             NaN  
123601             3.0       0.0             NaN  
29728              1.0       2.0             NaN  
26544              2.0       0.0             NaN  
54261              NaN       NaN             NaN  
124928             NaN       NaN             NaN  
171373             2.0       2.0             NaN  
32076             10.0       2.0             NaN  
194239           100.0       2.0             NaN  
41608             33.0       2.0             NaN  
60570              NaN       NaN             NaN  


In [8]:
data.iloc[:,33:43].sample(20)

Existing Construction Type Existing Construction Type Description  \
36554                          5.0                         wood frame (5)   
179109                         5.0                         wood frame (5)   
104632                         5.0                         wood frame (5)   
138592                         NaN                                    NaN   
65738                          NaN                                    NaN   
151325                         NaN                                    NaN   
103420                         5.0                         wood frame (5)   
87557                          5.0                         wood frame (5)   
112635                         5.0                         wood frame (5)   
57182                          1.0                          constr type 1   
99729                          NaN                                    NaN   
123376                         5.0                         wood frame (5)   
34858                          5.0                         wood frame (5)   
150061                         1.0                          constr type 1   
30332                          3.0                          constr type 3   
44879                          5.0                         wood frame (5)   
136680                         5.0                         wood frame (5)   
75450                          5.0                         wood frame (5)   
23776                          1.0                          constr type 1   
173641                         5.0                         wood frame (5)   

        Proposed Construction Type Proposed Construction Type Description  \
36554                          5.0                         wood frame (5)   
179109                         5.0                         wood frame (5)   
104632                         5.0                         wood frame (5)   
138592                         NaN                                    NaN   
65738                          NaN                                    NaN   
151325                         NaN                                    NaN   
103420                         5.0                         wood frame (5)   
87557                          5.0                         wood frame (5)   
112635                         5.0                         wood frame (5)   
57182                          1.0                          constr type 1   
99729                          NaN                                    NaN   
123376                         5.0                         wood frame (5)   
34858                          5.0                         wood frame (5)   
150061                         1.0                          constr type 1   
30332                          3.0                          constr type 3   
44879                          5.0                         wood frame (5)   
136680                         5.0                         wood frame (5)   
75450                          5.0                         wood frame (5)   
23776                          1.0                          constr type 1   
173641                         5.0                         wood frame (5)   

       Site Permit  Supervisor District Neighborhoods - Analysis Boundaries  \
36554          NaN                  7.0                        Inner Sunset   
179109         NaN                  6.0                     South of Market   
104632           Y                 10.0               Bayview Hunters Point   
138592         NaN                  9.0                      Bernal Heights   
65738          NaN                  2.0                    Presidio Heights   
151325         NaN                  5.0                        Hayes Valley   
103420         NaN                  2.0                     Pacific Heights   
87557          NaN                  3.0                           Chinatown   
112635         NaN                  8.0                      

Veamos los tipos de datos que tenemos

In [9]:
data_types = data.dtypes
data_types = pd.DataFrame(data_types)
data_types = data_types.rename(columns = {0:'ReadingDataType'})
data_types.index.names = ['FieldName']
data_types

ReadingDataType
FieldName                                             
Permit Number                                   object
Permit Type                                      int64
Permit Type Definition                          object
Permit Creation Date                            object
Block                                           object
Lot                                             object
Street Number                                    int64
Street Number Suffix                            object
Street Name                                     object
Street Suffix                                   object
Unit                                           float64
Unit Suffix                                     object
Description                                     object
Current Status                                  object
Current Status Date                             object
Filed Date                                      object
Issued Date                                     object
Completed Date                                  object
First Construction Document Date                object
Structural Notification                         object
Number of Existing Stories                     float64
Number of Proposed Stories                     float64
Voluntary Soft-Story Retrofit                   object
Fire Only Permit                                object
Permit Expiration Date                          object
Estimated Cost                                 float64
Revised Cost                                   float64
Existing Use                                    object
Existing Units                                 float64
Proposed Use                                    object
Proposed Units                                 float64
Plansets                                       float64
TIDF Compliance                                 object
Existing Construction Type                     float64
Existing Construction Type Description          object
Proposed Construction Type                     float64
Proposed Construction Type Description          object
Site Permit                                     object
Supervisor District                            float64
Neighborhoods - Analysis Boundaries             object
Zipcode                                        float64
Location                                        object
Record ID                                        int64

Según la metadata que acompaña el dataset, tenemos lo siguiente:

In [10]:
data_def_types = pd.DataFrame([["Permit Number","String"],
["Permit Type","Numeric"],
["Permit Type Definition","String"],
["Permit Creation Date","DateTime"],
["Block","String"],
["Lot","String"],
["Street Number","Numeric"],
["Street Number Suffix","String"],
["Street Name","String"],
["Street Suffix","String"],
["Unit","String"],
["Unit Suffix","String"],
["Description","String"],
["Current Status","String"],
["Current Status Date","DateTime"],
["Filed Date","DateTime"],
["Issued Date","String"],
["Completed Date","String"],
["First Construction Document Date","String"],
["Structural Notification","String"],
["Number of Existing Stories","String"],
["Number of Proposed Stories","String"],
["Voluntary Soft-Story Retrofit","String"],
["Fire Only Permit","String"],
["Permit Expiration Date","String"],
["Estimated Cost","String"],
["Revised Cost","String"],
["Existing Use","String"],
["Existing Units","String"],
["Proposed Use","String"],
["Proposed Units","String"],
["Plansets","String"],
["TIDF Compliance","String"],
["Existing Construction Type","String"],
["Existing Construction Type Description","String"],
["Proposed Construction Type","String"],
["Proposed Construction Type Description","String"],
["Site Permit","String"],
["Supervisor District","Numeric"],
["Neighborhoods - Analysis Boundaries","String"],
["Zipcode","Numeric"],
["Location","String"],
["Record ID","Numeric"]],
columns=["FieldName","ProvidedType"])

data_def_types = data_def_types.set_index("FieldName")
data_def_types

ProvidedType
FieldName                                          
Permit Number                                String
Permit Type                                 Numeric
Permit Type Definition                       String
Permit Creation Date                       DateTime
Block                                        String
Lot                                          String
Street Number                               Numeric
Street Number Suffix                         String
Street Name                                  String
Street Suffix                                String
Unit                                         String
Unit Suffix                                  String
Description                                  String
Current Status                               String
Current Status Date                        DateTime
Filed Date                                 DateTime
Issued Date                                  String
Completed Date                               String
First Construction Document Date             String
Structural Notification                      String
Number of Existing Stories                   String
Number of Proposed Stories                   String
Voluntary Soft-Story Retrofit                String
Fire Only Permit                             String
Permit Expiration Date                       String
Estimated Cost                               String
Revised Cost                                 String
Existing Use                                 String
Existing Units                               String
Proposed Use                                 String
Proposed Units                               String
Plansets                                     String
TIDF Compliance                              String
Existing Construction Type                   String
Existing Construction Type Description       String
Proposed Construction Type                   String
Proposed Construction Type Description       String
Site Permit                                  String
Supervisor District                         Numeric
Neighborhoods - Analysis Boundaries          String
Zipcode                                     Numeric
Location                                     String
Record ID                                   Numeric

Comparando, tenemos lo sigiente:

In [11]:
pd.merge(data_types, data_def_types, left_index=True, right_index=True)

ReadingDataType ProvidedType
FieldName                                                          
Permit Number                                   object       String
Permit Type                                      int64      Numeric
Permit Type Definition                          object       String
Permit Creation Date                            object     DateTime
Block                                           object       String
Lot                                             object       String
Street Number                                    int64      Numeric
Street Number Suffix                            object       String
Street Name                                     object       String
Street Suffix                                   object       String
Unit                                           float64       String
Unit Suffix                                     object       String
Description                                     object       String
Current Status                                  object       String
Current Status Date                             object     DateTime
Filed Date                                      object     DateTime
Issued Date                                     object       String
Completed Date                                  object       String
First Construction Document Date                object       String
Structural Notification                         object       String
Number of Existing Stories                     float64       String
Number of Proposed Stories                     float64       String
Voluntary Soft-Story Retrofit                   object       String
Fire Only Permit                                object       String
Permit Expiration Date                          object       String
Estimated Cost                                 float64       String
Revised Cost                                   float64       String
Existing Use                                    object       String
Existing Units                                 float64       String
Proposed Use                                    object       String
Proposed Units                                 float64       String
Plansets                                       float64       String
TIDF Compliance                                 object       String
Existing Construction Type                     float64       String
Existing Construction Type Description          object       String
Proposed Construction Type                     float64       String
Proposed Construction Type Description          object       String
Site Permit                                     object       String
Supervisor District                            float64      Numeric
Neighborhoods - Analysis Boundaries             object       String
Zipcode                                        float64      Numeric
Location                                        object       String
Record ID                                        int64      Numeric

Como puede verse en la comparación, algunos campos originalmente tipificados como `String` han sido identificados como numéricos, pero esto no generará inconvenientes. Además todos los campos numéricos han sido catalogados como `int` o `float` correctamente. Sin embargo, las columnas **Permit Creation Date**, **Current Status Date** y **Filed Date** han sido identificadas como `object`, pero contienen datos del tipo `DateTime`.

Para salvar este error de tipificación, realizamos nuevamente la lectura, especificando el tipo de datos para estas columnas:

In [12]:
data = pd.read_csv('Building_Permits.csv',low_memory=False,
                   parse_dates=["Permit Creation Date","Current Status Date","Filed Date"])

In [13]:
data.head(10)

Permit Number  Permit Type            Permit Type Definition  \
0  201505065519            4                      sign - erect   
1  201604195146            4                      sign - erect   
2  201605278609            3  additions alterations or repairs   
3  201611072166            8            otc alterations permit   
4  201611283529            6                       demolitions   
5  201706149344            8            otc alterations permit   
6  201706300814            8            otc alterations permit   
7       M803667            8            otc alterations permit   
8       M804227            8            otc alterations permit   
9       M804767            8            otc alterations permit   

  Permit Creation Date Block  Lot  Street Number Street Number Suffix  \
0           2015-05-06  0326  023            140                  NaN   
1           2016-04-19  0306  007            440                  NaN   
2           2016-05-27  0595  203           1647                  NaN   
3           2016-11-07  0156  011           1230                  NaN   
4           2016-11-28  0342  001            950                  NaN   
5           2017-06-14  4105  009            800                  NaN   
6           2017-06-30  1739  020           1291                  NaN   
7           2017-06-30  4789  014           1465                  NaN   
8           2017-07-05  1212  054           2094                  NaN   
9           2017-07-06  1259  016             89                  NaN   

  Street Name Street Suffix  Unit Unit Suffix  \
0       Ellis            St   NaN         NaN   
1       Geary            St   0.0         NaN   
2     Pacific            Av   NaN         NaN   
3     Pacific            Av   0.0         NaN   
4      Market            St   NaN         NaN   
5     Indiana            St   NaN         NaN   
6        11th            Av   0.0         NaN   
7      Revere            Av   NaN         NaN   
8        Fell            St   NaN         NaN   
9      Alpine            Tr   NaN         NaN   

                                         Description Current Status  \
0  ground fl facade: to erect illuminated, electr...        expired   
1            remove (e) awning and associated signs.         issued   
2                    installation of separating wall      withdrawn   
3           repair dryrot & stucco at front of bldg.       complete   
4  demolish retail/office/commercial 3-story buil...         issued   
5                                          evac maps         issued   
6                                         re-roofing       complete   
7                                        streetspace         issued   
8                                street space permit         issued   
9                                       street space         issued   

  Current Status Date Filed Date Issued Date Completed Date  \
0          2017-12-21 2015-05-06  11/09/2015            NaN   
1          2017-08-03 2016-04-19  08/03/2017            NaN   
2          2017-09-26 2016-05-27         NaN            NaN   
3          2017-07-24 2016-11-07  07/18/2017     07/24/2017   
4          2017-12-01 2016-11-28  12/01/2017            NaN   
5          2017-07-06 2017-06-14  07/06/2017            NaN   
6          2017-07-12 2017-06-30  06/30/2017     07/12/2017   
7          2017-06-30 2017-06-30  06/30/2017            NaN   
8          2017-07-05 2017-07-05  07/05/2017            NaN   
9          2017-07-06 2017-07-06  07/06/2017            NaN   

  First Construction Document Date Structural Notification  \
0                       11/09/2015                     NaN   
1                       08/03/2017                     NaN   
2                              NaN                     NaN   
3                       07/18/2017                     NaN   
4                       11/20/2017                     NaN   
5                       07/06/2017                     NaN   
6                    

In [14]:
data_types = data.dtypes
data_types = pd.DataFrame(data_types)
data_types = data_types.rename(columns = {0:'ReadingDataType'})
data_types.index.names = ['FieldName']
data_types

ReadingDataType
FieldName                                             
Permit Number                                   object
Permit Type                                      int64
Permit Type Definition                          object
Permit Creation Date                    datetime64[ns]
Block                                           object
Lot                                             object
Street Number                                    int64
Street Number Suffix                            object
Street Name                                     object
Street Suffix                                   object
Unit                                           float64
Unit Suffix                                     object
Description                                     object
Current Status                                  object
Current Status Date                     datetime64[ns]
Filed Date                              datetime64[ns]
Issued Date                                     object
Completed Date                                  object
First Construction Document Date                object
Structural Notification                         object
Number of Existing Stories                     float64
Number of Proposed Stories                     float64
Voluntary Soft-Story Retrofit                   object
Fire Only Permit                                object
Permit Expiration Date                          object
Estimated Cost                                 float64
Revised Cost                                   float64
Existing Use                                    object
Existing Units                                 float64
Proposed Use                                    object
Proposed Units                                 float64
Plansets                                       float64
TIDF Compliance                                 object
Existing Construction Type                     float64
Existing Construction Type Description          object
Proposed Construction Type                     float64
Proposed Construction Type Description          object
Site Permit                                     object
Supervisor District                            float64
Neighborhoods - Analysis Boundaries             object
Zipcode                                        float64
Location                                        object
Record ID                                        int64

Ahora sí, vemos que el tipo de dato de estas columnas es `DateTime`. De hecho, deberíamos poder calcular el rango de fechas de las creaciones de permisos, por ejemplo:

In [15]:
a = data['Permit Creation Date'].max() - data['Permit Creation Date'].min()
#a / pd.Timedelta(1,'M')
a

Timedelta('2158 days 00:00:00')

In [16]:
data.describe(include='all')

Permit Number    Permit Type  Permit Type Definition  \
count         198900  198900.000000                  198900   
unique        181495            NaN                       8   
top     201602179765            NaN  otc alterations permit   
freq             101            NaN                  178844   
first            NaN            NaN                     NaN   
last             NaN            NaN                     NaN   
mean             NaN       7.522323                     NaN   
std              NaN       1.457451                     NaN   
min              NaN       1.000000                     NaN   
25%              NaN       8.000000                     NaN   
50%              NaN       8.000000                     NaN   
75%              NaN       8.000000                     NaN   
max              NaN       8.000000                     NaN   

       Permit Creation Date   Block     Lot  Street Number  \
count                198900  198900  198900  198900.000000   
unique                 1291    4896    1055            NaN   
top     2017-09-15 00:00:00    3708     001            NaN   
freq                    413    1195   10114            NaN   
first   2012-03-28 00:00:00     NaN     NaN            NaN   
last    2018-02-23 00:00:00     NaN     NaN            NaN   
mean                    NaN     NaN     NaN    1121.728944   
std                     NaN     NaN     NaN    1135.768948   
min                     NaN     NaN     NaN       0.000000   
25%                     NaN     NaN     NaN     235.000000   
50%                     NaN     NaN     NaN     710.000000   
75%                     NaN     NaN     NaN    1700.000000   
max                     NaN     NaN     NaN    8400.000000   

       Street Number Suffix Street Name Street Suffix          Unit  \
count                  2216      198900        196132  29479.000000   
unique                   18        1704            21           NaN   
top                       A      Market            St           NaN   
freq                   1501        5443        138358           NaN   
first                   NaN         NaN           NaN           NaN   
last                    NaN         NaN           NaN           NaN   
mean                    NaN         NaN           NaN     78.517182   
std                     NaN         NaN           NaN    326.981324   
min                     NaN         NaN           NaN      0.000000   
25%                     NaN         NaN           NaN      0.000000   
50%                     NaN         NaN           NaN      0.000000   
75%                     NaN         NaN           NaN      1.000000   
max                     NaN         NaN           NaN   6004.000000   

       Unit Suffix   Description Current Status  Current Status Date  \
count         1961        198610         198900               198900   
unique         164        134272             14                 1307   
top              A  street space       complete  2017-09-15 00:00:00   
freq           346         22163          97077                  474   
first          NaN           NaN            NaN  2013-01-02 00:00:00   
last           NaN           NaN            NaN  2019-03-02 00:00:00   
mean           NaN           NaN            NaN                  NaN   
std            NaN           NaN            NaN                  NaN   
min            NaN           NaN            NaN                  NaN   
25%            NaN           NaN            NaN                  NaN   
50%            NaN           NaN            NaN                  NaN   
75%            NaN           NaN            NaN                  NaN   
max            NaN           NaN            NaN                  NaN   

                 Filed Date Issued Date Completed Date  \
count                198900      183960          97191   
unique                 1288        1289           1300   
top     2017-09-15 00:00:00  06/15/2016     10/30/2015   
freq               

Con esto, podemos decir que la importación se ha realizado correctamente. Podemos pasar al punto dos de la lista.

### 1.2. Asegurar la Existencia de IDs o Claves Únicas

El siguiente paso implica chequear que no existen datos duplicados y que las claves, si existen, son únicas.

En este sentido, se podría esperar que el `Permit Number` sea único. Sin embargo, del resumen previo se puede deducir que esto no es así, ya que la cuenta es bastante superior a la cantidad de valores únicos. Es decir, existen más de 17000 permisos con el mismo `Permit Number`.

In [29]:
data.shape

(198900, 43)

In [30]:
data_ind = data.set_index("Permit Number")

In [31]:
data_ind[data_ind.index.duplicated(keep=False)].sort_index()

Permit Type  Permit Type Definition Permit Creation Date Block  \
Permit Number                                                                   
201301027072             8  otc alterations permit           2013-01-02  3582   
201301027072             8  otc alterations permit           2013-01-02  3582   
201301027079             8  otc alterations permit           2013-01-02  1230   
201301027079             8  otc alterations permit           2013-01-02  1230   
201301027081             8  otc alterations permit           2013-01-02  3196   
201301027081             8  otc alterations permit           2013-01-02  3196   
201301027093             8  otc alterations permit           2013-01-02  1635   
201301027093             8  otc alterations permit           2013-01-02  1635   
201301027095             8  otc alterations permit           2013-01-02  3564   
201301027095             8  otc alterations permit           2013-01-02  3564   
201301027105             8  otc alterations permit           2013-01-02  1173   
201301027105             8  otc alterations permit           2013-01-02  1173   
201301037165             8  otc alterations permit           2013-01-03  0535   
201301037165             8  otc alterations permit           2013-01-03  0535   
201301037171             8  otc alterations permit           2013-01-03  0520   
201301037171             8  otc alterations permit           2013-01-03  0520   
201301037188             8  otc alterations permit           2013-01-03  3643   
201301037188             8  otc alterations permit           2013-01-03  3643   
201301037191             8  otc alterations permit           2013-01-03  0998   
201301037191             8  otc alterations permit           2013-01-03  0998   
201301037195             8  otc alterations permit           2013-01-03  0976   
201301037195             8  otc alterations permit           2013-01-03  0976   
201301037195             8  otc alterations permit           2013-01-03  0976   
201301037195             8  otc alterations permit           2013-01-03  0976   
201301037207             8  otc alterations permit           2013-01-03  3732   
201301037207             8  otc alterations permit           2013-01-03  3732   
201301047213             8  otc alterations permit           2013-01-04  0849   
201301047213             8  otc alterations permit           2013-01-04  0849   
201301047218             8  otc alterations permit           2013-01-04  3537   
201301047218             8  otc alterations permit           2013-01-04  3537   
...                    ...                     ...                  ...   ...   
M888788                  8  otc alterations permit           2018-02-13  5514   
M888788                  8  otc alterations permit           2018-02-13  5514   
M888807                  8  otc alterations permit           2018-02-13  4003   
M888807                  8  otc alterations permit           2018-02-13  4003   
M889507                  8  otc alterations permit           2018-02-14  0152   
M889507                  8  otc alterations permit           2018-02-14  0152   
M889749                  8  otc alterations permit           2018-02-15  8725   
M889749                  8  otc alterations permit           2018-02-15  8725   
M889887                  8  otc alterations permit           2018-02-15  0545   
M889887                  8  otc alterations permit           2018-02-15  0545   
M889947                  8  otc alterations permit           2018-02-15  1350   
M889947                  8  otc alterations permit           2018-02-15  1350   
M890327                  8  otc alterations permit           2018-02-16  0474   
M890327                  8  otc alterations permit           2018-02-16  0474   
M890607                  8  otc alterations permit           2018-02-16  3533   
M890607                  8  otc alterations permit           2018-02-16  3533   
M890987                  8  otc alterations pe

Vemos que, al menos en los registros visualizados, la diferencia parece estar principalmente en la columna `Record ID`. Por otra parte, en las explicaciones del dataset propuesto por el autor describe este campo como "algún ID, no útil para esto". Dado que no agrega información, podemos suprimir esta columna de nuestro set de datos:

In [98]:
data_ind.columns

data_ind = data_ind[['Permit Type', 'Permit Type Definition', 'Permit Creation Date',
       'Block', 'Lot', 'Street Number', 'Street Number Suffix', 'Street Name',
       'Street Suffix', 'Unit', 'Unit Suffix', 'Description', 'Current Status',
       'Current Status Date', 'Filed Date', 'Issued Date', 'Completed Date',
       'First Construction Document Date', 'Structural Notification',
       'Number of Existing Stories', 'Number of Proposed Stories',
       'Voluntary Soft-Story Retrofit', 'Fire Only Permit',
       'Permit Expiration Date', 'Estimated Cost', 'Revised Cost',
       'Existing Use', 'Existing Units', 'Proposed Use', 'Proposed Units',
       'Plansets', 'TIDF Compliance', 'Existing Construction Type',
       'Existing Construction Type Description', 'Proposed Construction Type',
       'Proposed Construction Type Description', 'Site Permit',
       'Supervisor District', 'Neighborhoods - Analysis Boundaries', 'Zipcode',
       'Location']]
data_ind.shape

(198900, 41)

Si borramos los duplicados, considerando los valores de todas las columnas restantes, tendremos lo siguiente:

In [173]:
data_ind_uni = data_ind.drop_duplicates(subset = ['Permit Type', 'Permit Type Definition',
        'Permit Creation Date','Block', 'Lot', 'Street Number', 'Street Number Suffix', 'Street Name',
       'Street Suffix', 'Unit', 'Unit Suffix', 'Description', 'Current Status',
       'Current Status Date', 'Filed Date', 'Issued Date', 'Completed Date',
       'First Construction Document Date', 'Structural Notification',
       'Number of Existing Stories', 'Number of Proposed Stories',
       'Voluntary Soft-Story Retrofit', 'Fire Only Permit',
       'Permit Expiration Date', 'Estimated Cost', 'Revised Cost',
       'Existing Use', 'Existing Units', 'Proposed Use', 'Proposed Units',
       'Plansets', 'TIDF Compliance', 'Existing Construction Type',
       'Existing Construction Type Description', 'Proposed Construction Type',
       'Proposed Construction Type Description', 'Site Permit',
       'Supervisor District', 'Neighborhoods - Analysis Boundaries', 'Zipcode',
       'Location'], inplace=False)

data_ind_uni.shape

(198627, 41)

Vemos que sólo se han eliminado 273 registros, por lo que aún permanecen `Permit Numbers` repetidos. Analicemos nuevamente los valores de los restantes campos:

In [174]:
data_ind_dupl = data_ind_uni[data_ind_uni.index.duplicated(keep=False)].sort_index()
# data_ind_dupl.to_csv("Output/analisis_permit_num_repetido.csv")
data_ind_dupl

Permit Type  Permit Type Definition Permit Creation Date Block  \
Permit Number                                                                   
201301027072             8  otc alterations permit           2013-01-02  3582   
201301027072             8  otc alterations permit           2013-01-02  3582   
201301027079             8  otc alterations permit           2013-01-02  1230   
201301027079             8  otc alterations permit           2013-01-02  1230   
201301027081             8  otc alterations permit           2013-01-02  3196   
201301027081             8  otc alterations permit           2013-01-02  3196   
201301027093             8  otc alterations permit           2013-01-02  1635   
201301027093             8  otc alterations permit           2013-01-02  1635   
201301027095             8  otc alterations permit           2013-01-02  3564   
201301027095             8  otc alterations permit           2013-01-02  3564   
201301027105             8  otc alterations permit           2013-01-02  1173   
201301027105             8  otc alterations permit           2013-01-02  1173   
201301037165             8  otc alterations permit           2013-01-03  0535   
201301037165             8  otc alterations permit           2013-01-03  0535   
201301037171             8  otc alterations permit           2013-01-03  0520   
201301037171             8  otc alterations permit           2013-01-03  0520   
201301037188             8  otc alterations permit           2013-01-03  3643   
201301037188             8  otc alterations permit           2013-01-03  3643   
201301037191             8  otc alterations permit           2013-01-03  0998   
201301037191             8  otc alterations permit           2013-01-03  0998   
201301037195             8  otc alterations permit           2013-01-03  0976   
201301037195             8  otc alterations permit           2013-01-03  0976   
201301037195             8  otc alterations permit           2013-01-03  0976   
201301037195             8  otc alterations permit           2013-01-03  0976   
201301037207             8  otc alterations permit           2013-01-03  3732   
201301037207             8  otc alterations permit           2013-01-03  3732   
201301047213             8  otc alterations permit           2013-01-04  0849   
201301047213             8  otc alterations permit           2013-01-04  0849   
201301047218             8  otc alterations permit           2013-01-04  3537   
201301047218             8  otc alterations permit           2013-01-04  3537   
...                    ...                     ...                  ...   ...   
M888788                  8  otc alterations permit           2018-02-13  5514   
M888788                  8  otc alterations permit           2018-02-13  5514   
M888807                  8  otc alterations permit           2018-02-13  4003   
M888807                  8  otc alterations permit           2018-02-13  4003   
M889507                  8  otc alterations permit           2018-02-14  0152   
M889507                  8  otc alterations permit           2018-02-14  0152   
M889749                  8  otc alterations permit           2018-02-15  8725   
M889749                  8  otc alterations permit           2018-02-15  8725   
M889887                  8  otc alterations permit           2018-02-15  0545   
M889887                  8  otc alterations permit           2018-02-15  0545   
M889947                  8  otc alterations permit           2018-02-15  1350   
M889947                  8  otc alterations permit           2018-02-15  1350   
M890327                  8  otc alterations permit           2018-02-16  0474   
M890327                  8  otc alterations permit           2018-02-16  0474   
M890607                  8  otc alterations permit           2018-02-16  3533   
M890607                  8  otc alterations permit           2018-02-16  3533   
M890987                  8  otc alterations pe

En muchos casos, se observa que tanto `Street Number` como `Lot` pueden tener diferentes valores para el mismo `Permit Number`. Esto es algo factible, ya que el mismo permiso pudo haber aplicado a diferentes construcciones. Construimos una clave que los incorpore:

In [175]:
data_ind_uni.reset_index(inplace = True)

In [176]:
data_ind_uni['ID'] = data_ind_uni['Permit Number'] + ' - ' + data_ind_uni['Lot'] + ' - ' + data_ind_uni['Street Number'].astype(str)
data_ind_uni.head(2)

C:\Users\Julieta\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


Permit Number  Permit Type Permit Type Definition Permit Creation Date  \
0  201505065519            4           sign - erect           2015-05-06   
1  201604195146            4           sign - erect           2016-04-19   

  Block  Lot  Street Number Street Number Suffix Street Name Street Suffix  \
0  0326  023            140                  NaN       Ellis            St   
1  0306  007            440                  NaN       Geary            St   

   Unit Unit Suffix                                        Description  \
0   NaN         NaN  ground fl facade: to erect illuminated, electr...   
1   0.0         NaN            remove (e) awning and associated signs.   

  Current Status Current Status Date Filed Date Issued Date Completed Date  \
0        expired          2017-12-21 2015-05-06  11/09/2015            NaN   
1         issued          2017-08-03 2016-04-19  08/03/2017            NaN   

  First Construction Document Date Structural Notification  \
0                       11/09/2015                     NaN   
1                       08/03/2017                     NaN   

   Number of Existing Stories  Number of Proposed Stories  \
0                         6.0                         NaN   
1                         7.0                         NaN   

  Voluntary Soft-Story Retrofit Fire Only Permit Permit Expiration Date  \
0                           NaN              NaN             11/03/2016   
1                           NaN              NaN             12/03/2017   

   Estimated Cost  Revised Cost         Existing Use  Existing Units  \
0          4000.0        4000.0  tourist hotel/motel           143.0   
1             1.0         500.0  tourist hotel/motel             NaN   

  Proposed Use  Proposed Units  Plansets TIDF Compliance  \
0          NaN             NaN       2.0             NaN   
1          NaN             NaN       2.0             NaN   

   Existing Construction Type Existing Construction Type Description  \
0                         3.0                          constr type 3   
1                         3.0                          constr type 3   

   Proposed Construction Type Proposed Construction Type Description  \
0                         NaN                                    NaN   
1                         NaN                                    NaN   

  Site Permit  Supervisor District Neighborhoods - Analysis Boundaries  \
0         NaN                  3.0                          Tenderloin   
1         NaN                  3.0                          Tenderloin   

   Zipcode                                   Location  \
0  94102.0  (37.785719256680785, -122.40852313194863)   
1  94102.0   (37.78733980600732, -122.41063199757738)   

                         ID  
0  201505065519 - 023 - 140  
1  201604195146 - 007 - 440

Una vez que obtenemos este ID, volvemos a preguntar por los valores duplicados:

In [177]:
data_ind_uni.set_index('ID', inplace = True)

In [185]:
data_ind_dupl = data_ind_uni[data_ind_uni.index.duplicated(keep=False)].sort_index()

data_ind_dupl.head(20)

Permit Number  Permit Type  \
ID                                                     
201302049435 - 062 - 771   201302049435            8   
201302049435 - 062 - 771   201302049435            8   
201302059475 - 098 - 511   201302059475            8   
201302059475 - 098 - 511   201302059475            8   
201302089862 - 009A - 234  201302089862            8   
201302089862 - 009A - 234  201302089862            8   
201302119974 - 016 - 436   201302119974            8   
201302119974 - 016 - 436   201302119974            8   
201303081867 - 016 - 2592  201303081867            8   
201303081867 - 016 - 2592  201303081867            8   
201303132112 - 040 - 194   201303132112            8   
201303132112 - 040 - 194   201303132112            8   
201303132112 - 040 - 196   201303132112            8   
201303132112 - 040 - 196   201303132112            8   
201303152346 - 020 - 1223  201303152346            8   
201303152346 - 020 - 1223  201303152346            8   
201303202663 - 014 - 2878  201303202663            8   
201303202663 - 014 - 2878  201303202663            8   
201304023644 - 047A - 62   201304023644            3   
201304023644 - 047A - 62   201304023644            3   

                                     Permit Type Definition  \
ID                                                            
201302049435 - 062 - 771             otc alterations permit   
201302049435 - 062 - 771             otc alterations permit   
201302059475 - 098 - 511             otc alterations permit   
201302059475 - 098 - 511             otc alterations permit   
201302089862 - 009A - 234            otc alterations permit   
201302089862 - 009A - 234            otc alterations permit   
201302119974 - 016 - 436             otc alterations permit   
201302119974 - 016 - 436             otc alterations permit   
201303081867 - 016 - 2592            otc alterations permit   
201303081867 - 016 - 2592            otc alterations permit   
201303132112 - 040 - 194             otc alterations permit   
201303132112 - 040 - 194             otc alterations permit   
201303132112 - 040 - 196             otc alterations permit   
201303132112 - 040 - 196             otc alterations permit   
201303152346 - 020 - 1223            otc alterations permit   
201303152346 - 020 - 1223            otc alterations permit   
201303202663 - 014 - 2878            otc alterations permit   
201303202663 - 014 - 2878            otc alterations permit   
201304023644 - 047A - 62   additions alterations or repairs   
201304023644 - 047A - 62   additions alterations or repairs   

                          Permit Creation Date Block   Lot  Street Number  \
ID                                                                          
201302049435 - 062 - 771            2013-02-04  3729   062            771   
201302049435 - 062 - 771            2013-02-04  3729   062            771   
201302059475 - 098 - 511            2013-02-05  3726   098            511   
201302059475 - 098 - 511            2013-02-05  3726   098            511   
201302089862 - 009A - 234           2013-02-08  0868  009A            234   
201302089862 - 009A - 234           2013-02-08  0868  009A            234   
201302119974 - 016 - 436            2013-02-11  1020   016            436   
201302119974 - 016 - 436            2013-02-11  1020   016            436   
201303081867 - 016 - 2592           2013-03-08  1053   016           2592   
201303081867 - 016 - 2592           2013-03-08  1053   016           2592   
201303132112 - 040 - 194            2013-03-13  6615   040            194   
201303132112 - 040 - 194            2013-03-13  6615   040            194   
201303132112 - 040 - 196            2013-03-13  6615   040            196   
201303132112 - 040 - 196            2013-03-13  6615   040            196   
201303152346 - 020 - 1223           2013-03-15  5654   020           1223   
201303152346 - 020 - 1223           2013-03-15  5654   020           1223   
201303202663 - 014 -

De 31489 `Permit Numbers` duplicados, pasamos a tan solo 893 registros repetidos. Además, vemos que la principal diferencia de valores entre `ID's` repetidos se produce en las columnas `Street Number Suffix`, `Unit` y `Unit Suffix`. Para asegurarnos que tome aquellos registros que no tienen NaN en estos campos, ordenamos inicialmente el dataset.

In [188]:
data_ind_uni.sort_values(by = ['Current Status Date','Street Number Suffix', 'Unit Suffix','Unit'], axis=0, ascending=False, inplace=True,
                         na_position='last')

data_ind_uni.tail(5)

C:\Users\Julieta\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


Permit Number  Permit Type  Permit Type Definition  \
ID                                                                            
M862628 - 017A - 1228          M862628            8  otc alterations permit   
201712055595 - 014 - 580  201712055595            8  otc alterations permit   
M863507 - 019 - 1568           M863507            8  otc alterations permit   
M863747 - 029 - 795            M863747            8  otc alterations permit   
M864287 - 006 - 838            M864287            8  otc alterations permit   

                         Permit Creation Date Block   Lot  Street Number  \
ID                                                                         
M862628 - 017A - 1228              2017-12-05  0113  017A           1228   
201712055595 - 014 - 580           2017-12-05  0271   014            580   
M863507 - 019 - 1568               2017-12-06  4318   019           1568   
M863747 - 029 - 795                2017-12-06  0298   029            795   
M864287 - 006 - 838                2017-12-07  0160   006            838   

                         Street Number Suffix Street Name Street Suffix  Unit  \
ID                                                                              
M862628 - 017A - 1228                     NaN  Montgomery            St   NaN   
201712055595 - 014 - 580                  NaN        Bush            St   NaN   
M863507 - 019 - 1568                      NaN     Indiana            St   NaN   
M863747 - 029 - 795                       NaN      Sutter            St   NaN   
M864287 - 006 - 838                       NaN     Pacific            Av   NaN   

                         Unit Suffix                          Description  \
ID                                                                          
M862628 - 017A - 1228            NaN                         street space   
201712055595 - 014 - 580         NaN  fire alarm upgrade ref 201704123852   
M863507 - 019 - 1568             NaN                         street space   
M863747 - 029 - 795              NaN                  street space permit   
M864287 - 006 - 838              NaN                  street space permit   

                         Current Status Current Status Date Filed Date  \
ID                                                                       
M862628 - 017A - 1228            issued          2017-12-05 2017-12-05   
201712055595 - 014 - 580         issued          2017-12-06 2017-12-05   
M863507 - 019 - 1568             issued          2017-12-06 2017-12-06   
M863747 - 029 - 795              issued          2017-12-06 2017-12-06   
M864287 - 006 - 838              issued          2017-12-07 2017-12-07   

                         Issued Date Completed Date  \
ID                                                    
M862628 - 017A - 1228     12/05/2017            NaN   
201712055595 - 014 - 580  12/06/2017            NaN   
M863507 - 019 - 1568      12/06/2017            NaN   
M863747 - 029 - 795       12/06/2017            NaN   
M864287 - 006 - 838       12/07/2017            NaN   

                         First Construction Document Date  \
ID                                                          
M862628 - 017A - 1228                          12/05/2017   
201712055595 - 014 - 580                       12/06/2017   
M863507 - 019 - 1568                           12/06/2017   
M863747 - 029 - 795                            12/06/2017   
M864287 - 006 - 838                            12/07/2017   

                         Structural Notification  Number of Existing Stories  \
ID                                                                             
M862628 - 017A - 1228                        NaN                         NaN   
201712055595 - 014 - 580                     NaN                         4.0   
M863507 - 019 - 1568                         NaN                         NaN   
M863747 - 029 - 795                          NaN                         NaN   
M864287 - 006 - 8

In [180]:
data_ind_uni2 = data_ind_uni.drop_duplicates(subset = ['Permit Number','Permit Type', 'Permit Type Definition',
        'Permit Creation Date','Block', 'Lot', 'Street Number', 'Street Name',
       'Street Suffix', 'Description', 'Current Status',
       'Current Status Date', 'Filed Date', 'Issued Date', 'Completed Date',
       'First Construction Document Date', 'Structural Notification',
       'Number of Existing Stories', 'Number of Proposed Stories',
       'Voluntary Soft-Story Retrofit', 'Fire Only Permit',
       'Permit Expiration Date', 'Estimated Cost', 'Revised Cost',
       'Existing Use', 'Existing Units', 'Proposed Use', 'Proposed Units',
       'Plansets', 'TIDF Compliance', 'Existing Construction Type',
       'Existing Construction Type Description', 'Proposed Construction Type',
       'Proposed Construction Type Description', 'Site Permit',
       'Supervisor District', 'Neighborhoods - Analysis Boundaries', 'Zipcode',
       'Location'])
#Not 'Street Number Suffix', 'Unit Suffix','Unit'
data_ind_uni2.shape

(198149, 42)

Se eliminaron 478 registros más. Repetimos el procedimiento para volver a analizar los duplicados remanentes.

In [181]:
data_ind_dupl = data_ind_uni2[data_ind_uni2.index.duplicated(keep=False)].sort_index()
data_ind_dupl.shape

(32, 42)

In [183]:
data_ind_dupl

Permit Number  Permit Type  \
ID                                                     
201505186604 - 031 - 258   201505186604            3   
201505186604 - 031 - 258   201505186604            3   
201607011383 - 010 - 274   201607011383            3   
201607011383 - 010 - 274   201607011383            3   
201607283565 - 036 - 62    201607283565            1   
201607283565 - 036 - 62    201607283565            1   
201701056574 - 012 - 1856  201701056574            3   
201701056574 - 012 - 1856  201701056574            3   
201704264979 - 008 - 700   201704264979            8   
201704264979 - 008 - 700   201704264979            8   
201704264980 - 008 - 700   201704264980            8   
201704264980 - 008 - 700   201704264980            8   
201706098896 - 028 - 245   201706098896            3   
201706098896 - 028 - 245   201706098896            3   
201708023666 - 062 - 451   201708023666            8   
201708023666 - 062 - 451   201708023666            8   
201708255934 - 010 - 271   201708255934            8   
201708255934 - 010 - 271   201708255934            8   
201710242112 - 050 - 1866  201710242112            8   
201710242112 - 050 - 1866  201710242112            8   
201710262282 - 008 - 700   201710262282            8   
201710262282 - 008 - 700   201710262282            8   
201712045358 - 003 - 815   201712045358            8   
201712045358 - 003 - 815   201712045358            8   
201712206974 - 010 - 271   201712206974            8   
201712206974 - 010 - 271   201712206974            8   
201801118405 - 002 - 1106  201801118405            8   
201801118405 - 002 - 1106  201801118405            8   
201802151465 - 001 - 1201  201802151465            8   
201802151465 - 001 - 1201  201802151465            8   
201802161482 - 006 - 3912  201802161482            8   
201802161482 - 006 - 3912  201802161482            8   

                                     Permit Type Definition  \
ID                                                            
201505186604 - 031 - 258   additions alterations or repairs   
201505186604 - 031 - 258   additions alterations or repairs   
201607011383 - 010 - 274   additions alterations or repairs   
201607011383 - 010 - 274   additions alterations or repairs   
201607283565 - 036 - 62                    new construction   
201607283565 - 036 - 62                    new construction   
201701056574 - 012 - 1856  additions alterations or repairs   
201701056574 - 012 - 1856  additions alterations or repairs   
201704264979 - 008 - 700             otc alterations permit   
201704264979 - 008 - 700             otc alterations permit   
201704264980 - 008 - 700             otc alterations permit   
201704264980 - 008 - 700             otc alterations permit   
201706098896 - 028 - 245   additions alterations or repairs   
201706098896 - 028 - 245   additions alterations or repairs   
201708023666 - 062 - 451             otc alterations permit   
201708023666 - 062 - 451             otc alterations permit   
201708255934 - 010 - 271             otc alterations permit   
201708255934 - 010 - 271             otc alterations permit   
201710242112 - 050 - 1866            otc alterations permit   
201710242112 - 050 - 1866            otc alterations permit   
201710262282 - 008 - 700             otc alterations permit   
201710262282 - 008 - 700             otc alterations permit   
201712045358 - 003 - 815             otc alterations permit   
201712045358 - 003 - 815             otc alterations permit   
201712206974 - 010 - 271             otc alterations permit   
201712206974 - 010 - 271             otc alterations permit   
201801118405 - 002 - 1106            otc alterations permit   
201801118405 - 002 - 1106            otc alterations permit   
201802151465 - 001 - 1201            otc alterations permit   
201802151465 - 001 - 1201            otc alterations permit   
201802161482 - 006 - 3912            otc alterations permit   
201802161482 - 006 - 3912            otc a

Vemos que permanecen 32 registros repetidos aún. En muchos de los casos, la diferencia en los valores se produce en las columnas `Street Name` y `Street Suffix` así también como en los campos de fechas. En el caso del nombre de la calle y sufijo, probablemente se trate de construcciones en esquinas de manzana, que dan a dos calles. Pero como ya existe el dato de localización con coordenadas, es indistinto que mantengamos una calle u otra.

In [196]:
data_ind_uni2.sort_values(by = ['Current Status Date','Street Number Suffix', 'Unit Suffix','Unit', 'Street Suffix', 'Street Name','Issued Date','First Construction Document Date',
                                'Permit Expiration Date'], axis=0, ascending=False, inplace=True, na_position='last')

data_ind_uni2.head(5)

C:\Users\Julieta\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


Permit Number  Permit Type  \
ID                                                      
201611293622 - 005D - 1391  201611293622            8   
201802151417 - 121 - 1804   201802151417            8   
201802232089 - 038 - 1977   201802232089            8   
201504153737 - 049 - 240    201504153737            3   
201802232123 - 080 - 305    201802232123            8   

                                      Permit Type Definition  \
ID                                                             
201611293622 - 005D - 1391            otc alterations permit   
201802151417 - 121 - 1804             otc alterations permit   
201802232089 - 038 - 1977             otc alterations permit   
201504153737 - 049 - 240    additions alterations or repairs   
201802232123 - 080 - 305              otc alterations permit   

                           Permit Creation Date Block   Lot  Street Number  \
ID                                                                           
201611293622 - 005D - 1391           2016-11-29  0480  005D           1391   
201802151417 - 121 - 1804            2018-02-15  0728   121           1804   
201802232089 - 038 - 1977            2018-02-23  0542   038           1977   
201504153737 - 049 - 240             2015-04-15  6685   049            240   
201802232123 - 080 - 305             2018-02-23  6655   080            305   

                           Street Number Suffix Street Name Street Suffix  \
ID                                                                          
201611293622 - 005D - 1391                  NaN   Francisco            St   
201802151417 - 121 - 1804                     C       Ellis            St   
201802232089 - 038 - 1977                     B       Union            St   
201504153737 - 049 - 240                      A     Chenery            St   
201802232123 - 080 - 305                      A        30th            St   

                            Unit Unit Suffix  \
ID                                             
201611293622 - 005D - 1391   NaN         NaN   
201802151417 - 121 - 1804   68.0         NaN   
201802232089 - 038 - 1977    NaN         NaN   
201504153737 - 049 - 240     NaN         NaN   
201802232123 - 080 - 305     NaN         NaN   

                                                                  Description  \
ID                                                                              
201611293622 - 005D - 1391  kitchen: reface cabinets, new sinks, faucet an...   
201802151417 - 121 - 1804   unit 68 - remove & replace damage members of s...   
201802232089 - 038 - 1977   obtain final inspection for work approved unde...   
201504153737 - 049 - 240    (rear bldg) renovation to an (e) fire damaged ...   
201802232123 - 080 - 305    new door, bring stairs to code, adjust rails, ...   

                           Current Status Current Status Date Filed Date  \
ID                                                                         
201611293622 - 005D - 1391      cancelled          2019-03-02 2016-11-29   
201802151417 - 121 - 1804          issued          2018-02-23 2018-02-15   
201802232089 - 038 - 1977          issued          2018-02-23 2018-02-23   
201504153737 - 049 - 240           issued          2018-02-23 2015-04-15   
201802232123 - 080 - 305           issued          2018-02-23 2018-02-23   

                           Issued Date Completed Date  \
ID                                                      
201611293622 - 005D - 1391  11/29/2016            NaN   
201802151417 - 121 - 1804   02/23/2018            NaN   
201802232089 - 038 - 1977   02/23/2018            NaN   
201504153737 - 049 - 240    02/23/2018            NaN   
201802232123 - 080 - 305    02/23/2018            NaN   

                           First Construction Document Date  \
ID                                                            
201611293622 - 005D - 1391                       11/29/2016   
201802151417 - 121 - 1804                        02/23/2018   
2

In [191]:
data_ind_uni3 = data_ind_uni2.drop_duplicates(subset = ['Permit Number','Permit Type', 'Permit Type Definition', 
       'Block', 'Lot', 'Street Number', 'Description', 'Current Status',
       'Current Status Date', 'Filed Date', 'Issued Date', 'Completed Date', 'Structural Notification',
       'Number of Existing Stories', 'Number of Proposed Stories',
       'Voluntary Soft-Story Retrofit', 'Fire Only Permit', 'Estimated Cost', 'Revised Cost',
       'Existing Use', 'Existing Units', 'Proposed Use', 'Proposed Units',
       'Plansets', 'TIDF Compliance', 'Existing Construction Type',
       'Existing Construction Type Description', 'Proposed Construction Type',
       'Proposed Construction Type Description', 'Site Permit',
       'Supervisor District', 'Neighborhoods - Analysis Boundaries', 'Zipcode',
       'Location'])
#Not 'Street Number Suffix', 'Unit Suffix','Unit', 'Permit Creation Date','Street Name'
#,'Street Suffix', 'First Construction Document Date', 'Permit Expiration Date'

data_ind_uni3.shape

(198142, 42)

Se eliminaron 7 registros más. Chequeemos nuevamente duplicados:

In [192]:
data_ind_dupl = data_ind_uni3[data_ind_uni3.index.duplicated(keep=False)].sort_index()
data_ind_dupl.shape

(24, 42)

In [193]:
data_ind_dupl

Permit Number  Permit Type  \
ID                                                     
201505186604 - 031 - 258   201505186604            3   
201505186604 - 031 - 258   201505186604            3   
201607011383 - 010 - 274   201607011383            3   
201607011383 - 010 - 274   201607011383            3   
201607283565 - 036 - 62    201607283565            1   
201607283565 - 036 - 62    201607283565            1   
201701056574 - 012 - 1856  201701056574            3   
201701056574 - 012 - 1856  201701056574            3   
201706098896 - 028 - 245   201706098896            3   
201706098896 - 028 - 245   201706098896            3   
201708255934 - 010 - 271   201708255934            8   
201708255934 - 010 - 271   201708255934            8   
201710242112 - 050 - 1866  201710242112            8   
201710242112 - 050 - 1866  201710242112            8   
201712045358 - 003 - 815   201712045358            8   
201712045358 - 003 - 815   201712045358            8   
201712206974 - 010 - 271   201712206974            8   
201712206974 - 010 - 271   201712206974            8   
201801118405 - 002 - 1106  201801118405            8   
201801118405 - 002 - 1106  201801118405            8   
201802151465 - 001 - 1201  201802151465            8   
201802151465 - 001 - 1201  201802151465            8   
201802161482 - 006 - 3912  201802161482            8   
201802161482 - 006 - 3912  201802161482            8   

                                     Permit Type Definition  \
ID                                                            
201505186604 - 031 - 258   additions alterations or repairs   
201505186604 - 031 - 258   additions alterations or repairs   
201607011383 - 010 - 274   additions alterations or repairs   
201607011383 - 010 - 274   additions alterations or repairs   
201607283565 - 036 - 62                    new construction   
201607283565 - 036 - 62                    new construction   
201701056574 - 012 - 1856  additions alterations or repairs   
201701056574 - 012 - 1856  additions alterations or repairs   
201706098896 - 028 - 245   additions alterations or repairs   
201706098896 - 028 - 245   additions alterations or repairs   
201708255934 - 010 - 271             otc alterations permit   
201708255934 - 010 - 271             otc alterations permit   
201710242112 - 050 - 1866            otc alterations permit   
201710242112 - 050 - 1866            otc alterations permit   
201712045358 - 003 - 815             otc alterations permit   
201712045358 - 003 - 815             otc alterations permit   
201712206974 - 010 - 271             otc alterations permit   
201712206974 - 010 - 271             otc alterations permit   
201801118405 - 002 - 1106            otc alterations permit   
201801118405 - 002 - 1106            otc alterations permit   
201802151465 - 001 - 1201            otc alterations permit   
201802151465 - 001 - 1201            otc alterations permit   
201802161482 - 006 - 3912            otc alterations permit   
201802161482 - 006 - 3912            otc alterations permit   

                          Permit Creation Date Block  Lot  Street Number  \
ID                                                                         
201505186604 - 031 - 258            2015-05-18  1383  031            258   
201505186604 - 031 - 258            2015-05-18  1383  031            258   
201607011383 - 010 - 274            2016-07-01  2698  010            274   
201607011383 - 010 - 274            2016-07-01  2698  010            274   
201607283565 - 036 - 62             2016-07-28  7120  036             62   
201607283565 - 036 - 62             2016-07-28  7120  036             62   
201701056574 - 012 - 1856           2017-01-05  2064  012           1856   
201701056574 - 012 - 1856           2017-01-05  2064  012           1856   
201706098896 - 028 - 245            2017-06-09  2697  028            245   
201706098896 - 028 - 245            2017-06-09  2697  028            245   
201708255934 - 010

Sigue habiendo algunos IDs (como, por ejemplo, 201712206974 - 010 - 271) que difieren en muchos de los conceptos. Por lo tanto, además de las fechas ya excluidas, también deberemos excluir (luego de haber ordenado previamente por estos conceptos) `Estimated Cost`, `Revised Cost`, `Existing Use`, `Existing Units`, `Proposed Use`, `Proposed Units`. El criterio adoptado fue quedarse con el registro cuya fecha de estado actual (`Current Status Date`) sea más reciente.

In [197]:
data_ind_uni3.sort_values(by = ['Current Status Date','Street Number Suffix', 'Unit Suffix','Unit', 'Street Suffix', 'Street Name','Issued Date','First Construction Document Date',
                                'Permit Expiration Date'], axis=0, ascending=False, inplace=True, na_position='last')

data_ind_uni3.head(5)

C:\Users\Julieta\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


Permit Number  Permit Type  \
ID                                                      
201611293622 - 005D - 1391  201611293622            8   
201802151417 - 121 - 1804   201802151417            8   
201802232089 - 038 - 1977   201802232089            8   
201504153737 - 049 - 240    201504153737            3   
201802232123 - 080 - 305    201802232123            8   

                                      Permit Type Definition  \
ID                                                             
201611293622 - 005D - 1391            otc alterations permit   
201802151417 - 121 - 1804             otc alterations permit   
201802232089 - 038 - 1977             otc alterations permit   
201504153737 - 049 - 240    additions alterations or repairs   
201802232123 - 080 - 305              otc alterations permit   

                           Permit Creation Date Block   Lot  Street Number  \
ID                                                                           
201611293622 - 005D - 1391           2016-11-29  0480  005D           1391   
201802151417 - 121 - 1804            2018-02-15  0728   121           1804   
201802232089 - 038 - 1977            2018-02-23  0542   038           1977   
201504153737 - 049 - 240             2015-04-15  6685   049            240   
201802232123 - 080 - 305             2018-02-23  6655   080            305   

                           Street Number Suffix Street Name Street Suffix  \
ID                                                                          
201611293622 - 005D - 1391                  NaN   Francisco            St   
201802151417 - 121 - 1804                     C       Ellis            St   
201802232089 - 038 - 1977                     B       Union            St   
201504153737 - 049 - 240                      A     Chenery            St   
201802232123 - 080 - 305                      A        30th            St   

                            Unit Unit Suffix  \
ID                                             
201611293622 - 005D - 1391   NaN         NaN   
201802151417 - 121 - 1804   68.0         NaN   
201802232089 - 038 - 1977    NaN         NaN   
201504153737 - 049 - 240     NaN         NaN   
201802232123 - 080 - 305     NaN         NaN   

                                                                  Description  \
ID                                                                              
201611293622 - 005D - 1391  kitchen: reface cabinets, new sinks, faucet an...   
201802151417 - 121 - 1804   unit 68 - remove & replace damage members of s...   
201802232089 - 038 - 1977   obtain final inspection for work approved unde...   
201504153737 - 049 - 240    (rear bldg) renovation to an (e) fire damaged ...   
201802232123 - 080 - 305    new door, bring stairs to code, adjust rails, ...   

                           Current Status Current Status Date Filed Date  \
ID                                                                         
201611293622 - 005D - 1391      cancelled          2019-03-02 2016-11-29   
201802151417 - 121 - 1804          issued          2018-02-23 2018-02-15   
201802232089 - 038 - 1977          issued          2018-02-23 2018-02-23   
201504153737 - 049 - 240           issued          2018-02-23 2015-04-15   
201802232123 - 080 - 305           issued          2018-02-23 2018-02-23   

                           Issued Date Completed Date  \
ID                                                      
201611293622 - 005D - 1391  11/29/2016            NaN   
201802151417 - 121 - 1804   02/23/2018            NaN   
201802232089 - 038 - 1977   02/23/2018            NaN   
201504153737 - 049 - 240    02/23/2018            NaN   
201802232123 - 080 - 305    02/23/2018            NaN   

                           First Construction Document Date  \
ID                                                            
201611293622 - 005D - 1391                       11/29/2016   
201802151417 - 121 - 1804                        02/23/2018   
2

In [202]:
data_ind_uni4 = data_ind_uni3.drop_duplicates(subset = ['Permit Number','Permit Type', 'Permit Type Definition', 
       'Block', 'Lot', 'Street Number','Current Status Date'])
#Not 'Street Number Suffix', 'Unit Suffix','Unit', 'Permit Creation Date','Street Name'
#,'Street Suffix', 'First Construction Document Date', 'Permit Expiration Date', 'Estimated Cost', 'Revised Cost',
#'Existing Use', 'Existing Units', 'Proposed Use', 'Proposed Units',

data_ind_uni4.shape

(198142, 42)

In [200]:
data_ind_dupl = data_ind_uni4[data_ind_uni4.index.duplicated(keep=False)].sort_index()
data_ind_dupl.shape

(22, 42)

Luego de explorar el dataset, además de las fechas tipificadas en la metadata, encontramos nuevos campos que contienen fechas: `Issued Date`, `First Construction Document Date` y `Permit Expiration Date`. Vamos a redefinir estos tipos de datos.

### 1.3. Despersonalizar Datos y Guardarlos en un Nuevo Archivo

En este caso, no es necesario despersonalizar ningún dato, dado que no existe información personal sensible.

### 1.4. Nunca Modificar los Datos Crudos u Originales

Al finalizar la limpieza, guardaremos nuevamente el dataset, para asegurarnos de no modificar los datos originales.

## 2. Pasos de Limpieza Necesarios

### 2.1. Etiquetas de Variables/Columnas y Problemas de Codificación/Encoding

### 2.2. Tratamiento de Valores Faltantes

Quitar o imputar

### 2.3. Codificación de Variables Categóricas

Las variables categóricas deben ser etiquetadas como variables numéricas, no como cadenas.

### 2.4. No Cambiar los Nombres de las Variables de la Fuente de Origen

### 2.5. Verificación de Consistencia de Datos

Aplicar reglas de integridad

### 2.6. Identificar y Documentar Valores Atípicos/Outliers

Calcular estadísticos

### 2.7. Evaluar Cómo Comprimir los Datos Para su Almacenamiento Más Eficiente

### 2.8. Guardar el Set de Datos con un Nombre Informativo

## 3. Pasos de Limpieza Deseables

### 3.1. Ordenar Variables/Columnas

Si es posible – primero ID, luego en el mismo orden que la fuente

### 3.2. Quitar Variables/Columnas Irrelevantes

que no Tienen información a analizar

### 3.3. Renombrar Variables de Grillas

### 3.4. Categorizar Resultados en “Otros”

Si tiene un campo de texto libre asociado, codificar en nuevos valores de la variable
categórica asociada. Revisar fuzzyness.

### 3.5. Agregar Metadata a los Datos

Cuando y como fueron obtenidos, limpieza realizada, asunciones, etc. Vincular con etiquetas del código fuente y los datos. Al menos incluir un README.